src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -517,135 +168,287 @@ namespace cublog
+  }
+
+  inline log_lsa
+  redo_parallel::redo_task::get_not_applied_log_lsa ()
+  {
+    assert (m_do_monitor_not_applied_log_lsa);
+
+    return m_not_applied_log_lsa.load ();
+  }
+
   /*********************************************************************
    * redo_parallel - definition
    *********************************************************************/
 
-  redo_parallel::redo_parallel (unsigned a_worker_count, minimum_log_lsa_monitor *a_minimum_log_lsa,
+  redo_parallel::redo_parallel (unsigned a_task_count, bool a_do_monitor_minimum_not_applied_log_lsa,
 				const log_rv_redo_context &copy_context)
-    : m_worker_pool (nullptr)
-    , m_job_queue { a_minimum_log_lsa }
+    : m_task_count { a_task_count }
+    , m_do_monitor_minimum_log_lsa { a_do_monitor_minimum_not_applied_log_lsa }
+    , m_worker_pool (nullptr)
     , m_waited_for_termination (false)
+    , m_adding_finished { false }
+    , m_outer_not_applied_log_lsa { MAX_LSA }
+    , m_calculated_minimum_not_applied_log_lsa { MAX_LSA }
+    , m_terminate_minimum_not_applied_log_lsa_calculation { false }
   {
-    assert (a_worker_count > 0);
+    assert (a_task_count > 0);
 
     const thread_type tt = log_is_in_crash_recovery () ? TT_RECOVERY : TT_REPLICATION;
     m_pool_context_manager = std::make_unique<cubthread::system_worker_entry_manager> (tt);
 
-    do_init_worker_pool (a_worker_count);
-    do_init_tasks (a_worker_count, copy_context);
+    do_init_worker_pool (a_task_count);
+    do_init_tasks (a_task_count, a_do_monitor_minimum_not_applied_log_lsa, copy_context);
+
+    if (m_do_monitor_minimum_log_lsa)
+      {
+	m_calculate_minimum_not_applied_log_lsa_thread = std::thread
+	{
+	  std::bind (&redo_parallel::redo_parallel::calculate_and_save_minimum_not_applied_log_lsa, std::ref (*this))
+	};
+      }
   }
 
   redo_parallel::~redo_parallel ()
   {
-    assert (m_job_queue.get_adding_finished ());
+    assert (m_adding_finished.load ());
     assert (m_worker_pool == nullptr);
     assert (m_waited_for_termination);
+    for (auto &redo_task: m_redo_tasks)
+      {
+	assert (redo_task->is_idle ());
+      }
+
+    if (m_do_monitor_minimum_log_lsa)
+      {
+	assert (m_calculate_minimum_not_applied_log_lsa_thread.joinable ());
+	m_terminate_minimum_not_applied_log_lsa_calculation = true;
+	m_calculate_minimum_not_applied_log_lsa_cv.notify_one ();
+	m_calculate_minimum_not_applied_log_lsa_thread.join ();
+      }
+    else
+      {
+	assert (!m_calculate_minimum_not_applied_log_lsa_thread.joinable ());
+	assert (m_calculated_minimum_not_applied_log_lsa == MAX_LSA);
+      }
   }
 
   void
   redo_parallel::add (redo_job_base *a_job)
   {
     assert (false == m_waited_for_termination);
-    assert (false == m_job_queue.get_adding_finished ());
+    assert (false == m_adding_finished.load ());
+
+    const vpid &job_vpid = a_job->get_vpid ();
+    const std::size_t vpid_hash = m_vpid_hash (job_vpid);
+    const std::size_t task_index = vpid_hash % m_task_count;
```

Comment:
The successions of operations is easier to be read in a single line:
```suggestion
    const std::size_t task_index = m_vpid_hash (a_job->get_vpid ()) % m_task_count;
```

Reply:
Done